test/Test.re

```diff
@@ -1,1061 +1,716 @@
+        let (rendered1, _) as actual =
+          TestRenderer.update(
+            rendered0,
+            <ChangeCounter label="defaultText" />
+          );
+        assertUpdate(
+          (
+            [
+              TestComponents.(
+                <ChangeCounter id=1 label="defaultText" counter=10 />
+              )
+            ],
+            None
+          ),
+          actual
+        );
+        let (rendered2, _) as actual2 =
+          TestRenderer.update(
+            rendered1,
+            <ChangeCounter label="updatedText" />
+          );
+        assertUpdate(
+          TestComponents.(
+            [<ChangeCounter id=1 label="updatedText" counter=11 />],
+            Some(
+              TestRenderer.{
+                subtreeChange: `Nested,
+                updateLog:
+                  ref([
+                    UpdateInstance({
+                      stateChanged: true,
+                      subTreeChanged: `NoChange,
+                      oldInstance:
+                        <ChangeCounter id=1 label="defaultText" counter=10 />,
+                      newInstance:
+                        <ChangeCounter id=1 label="updatedText" counter=11 />
+                    })
+                  ])
+              }
+            )
+          ),
+          actual2
+        );
+        let (rendered2f, _) as actual2f =
+          ReasonReact.RenderedElement.flushPendingUpdates(rendered2);
+        assertFlushUpdate(
+          TestComponents.(
+            [<ChangeCounter id=1 label="updatedText" counter=2011 />],
+            [
+              UpdateInstance({
+                stateChanged: true,
+                subTreeChanged: `NoChange,
+                oldInstance:
+                  <ChangeCounter id=1 label="updatedText" counter=11 />,
+                newInstance:
+                  <ChangeCounter id=1 label="updatedText" counter=2011 />
+              })
+            ]
+          ),
+          actual2f
+        );
+        let (rendered2f_mem, _) =
+          ReasonReact.RenderedElement.flushPendingUpdates(rendered2f);
+        check(
+          Alcotest.bool,
+          "it is memoized",
+          rendered2f_mem === rendered2f,
+          true
+        );
+        let (rendered2f_mem, _) =
+          ReasonReact.RenderedElement.flushPendingUpdates(rendered2f_mem);
+        check(
+          Alcotest.bool,
+          "it is memoized",
+          rendered2f_mem === rendered2f,
+          true
+        );
+        let (rendered3, _) as actual3 =
+          TestRenderer.update(
+            rendered2f_mem,
+            <ButtonWrapperWrapper wrappedText="updatedText" />
+          );
+        assertUpdate(
+          ~label="Updating components: ChangeCounter to ButtonWrapperWrapper",
+          TestComponents.(
+            [
+              <ButtonWrapperWrapper id=2 nestedText="wrappedText:updatedText" />
+            ],
+            Some(
+              TestRenderer.{
+                subtreeChange: `Nested,
```

Comment:
This doesn't make sense. If this caused ``` `Nested ```, what triggers `ReplaceElements` for the top level update?

Reply:
If I'm understanding this correctly, `ReplaceElements` is triggered only when we're replacing children element with a list of new elements (or inserting a list of new elements into the empty subtree).

I suggest we do a complete overhaul of `UpdateLog` types and outline all possible scenarios how updates could happen. (we could reuse our existing test suites establishing the correct expected output and work towards making the actual output match)

Reply:
I applaud this suggestion.